File: solutions/elixir/all-your-base/1/lib/all_your_base.ex
defmodule AllYourBase do
  @moduledoc """
  this module converts from one base to another
  """
  @doc """
  Given a number in input base, represented as a sequence of digits, converts it to output base,
  or returns an error tuple if either of the bases are less than 2
  """

  @spec convert(list, integer, integer) :: {:ok, list} | {:error, String.t()}
  def convert(_, _, output_base) when output_base < 2, do: {:error, "output base must be >= 2"}
  def convert(_, input_base, _) when input_base < 2, do: {:error, "input base must be >= 2"}
  def convert([], _input_base, _output_base), do: {:ok, [0]}
  def convert(digits, input_base, output_base) do
    if digits |> Enum.any?(&(&1 < 0 or &1 >= input_base)) do
      {:error, "all digits must be >= 0 and < input base"}
    else
      digits =
        digits
        |> to_int(input_base)
        |> from_int(output_base)

      {:ok, digits}
    end
  end

  @spec to_int(list, integer) :: integer
  defp to_int(digits, base) do
    Enum.reduce(digits, 0, fn digit, acc ->
      acc * base + digit
    end)
  end

  @spec from_int(integer, integer, list) :: list
  defp from_int(number, base, digits \\ [])
  defp from_int(0, _base, []), do: [0]
  defp from_int(0, _base, digits), do: digits
  defp from_int(number, base, digits) do
    digit = rem(number, base)
    number = trunc(number / base)
    from_int(number, base, [digit | digits])
  end
end

File: solutions/erlang/all-your-base/1/src/all_your_base.erl
-module(all_your_base).

-export([rebase/3]).

rebase(_digits, inputBase, _outputBase) when inputBase < 2 ->
  {error, "input base must be >= 2"};
rebase(_digits, _inputBase, outputBase) when outputBase < 2 ->
  {error, "output base must be >= 2"};
rebase(digits, inputBase, outputBase) ->
  case convert_to_base_10(digits, inputBase, length(digits) - 1, 0) of
    {error, Message} -> {error, Message};
    0 -> {ok, [0]};
    Number ->  convert_to_base(Number, outputBase, [])
  end.

convert_to_base_10([], _base, _n, sum) ->
  sum;
convert_to_base_10([h|_], base, _n, _sum) when h < 0 orelse h >= base ->
  {error , "all digits must satisfy 0 <= d < input base"};
convert_to_base_10([h|t], base, n, sum) ->
  convert_to_base_10(t, base, n - 1, sum + trunc(h * math:pow(base, n))).

convert_to_base(0, _base, acc) -> {ok, acc};
convert_to_base(n, base, acc) -> convert_to_base(n div base, base, [n rem base | acc]).